Composer/packages/extension-client/src/extensionEventEmitter.ts

```diff
@@ -0,0 +1,62 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+type EventListenerArgs<Args extends Record<string, unknown> = Record<string, unknown>> = Args;
+
+type EventListener<T extends Record<string, unknown> = Record<string, unknown>> = (
+  args: EventListenerArgs<T>
+) => Promise<void> | void;
+
+type Disposable = {
+  dispose: () => void;
+};
+
+class ComposerEventEmitter {
+  private static listeners: Record<string, EventListener[]> = {};
+
+  public static addListener(event: string, listener: EventListener): Disposable {
+    if (!this.listeners[event]) {
+      this.listeners[event] = [];
+    }
+
+    this.listeners[event].push(listener);
+
+    return {
+      dispose: () => this.removeListener(event, listener),
+    };
+  }
+
+  public static removeListener(event: string, listener: EventListener) {
+    const eventListeners = this.listeners[event];
+
+    if (eventListeners && eventListeners.length > 0) {
+      this.listeners[event] = eventListeners.filter((l) => l !== listener);
+    }
+  }
+
+  public static on(event: string, listener: EventListener): Disposable {
+    return this.addListener(event, listener);
+  }
+
+  public static async emit(event: string, args: EventListenerArgs) {
+    const listeners = this.listeners[event] ?? [];
+
+    for (const l of listeners) {
+      try {
+        await l(args);
+      } catch (err) {
+        console.error(err);
+      }
+    }
+  }
+
+  on(event: string, listener: EventListener<any>): Disposable {
+    return ComposerEventEmitter.addListener(event, listener);
```

Comment:
Would this be a temporary solution @yeze322 . Essentially, we can have just 1 listener for an event in the current pattern.